### ISSUE TITLE
[Bug]: multi GPU crashes backend

### ISSUE THREAD

**Issue by mrseeker** [PygmalionAI/aphrodite-engine#359]
### Your current environment

```
Docker container, can't check
```

### 🐛 Describe the bug

When I set `NUM_GPUS` to 8 (due to having a server with 8 GPU's) I get the following error (sorry but the system hates to properly log errors):
```
ERROR:    Exception in ASGI application
--
ERROR:    Traceback (most recent call last):
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 442, in engine_step
ERROR:        request_outputs = await self.engine.step_async()
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 208, in step_async
ERROR:        all_outputs = await self._run_workers_async(
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 293, in _run_workers_async
ERROR:        all_outputs = await asyncio.gather(*coros)
ERROR:    asyncio.exceptions.CancelledError
ERROR:
ERROR:    During handling of the above exception, another exception occurred:
ERROR:
ERROR:    Traceback (most recent call last):
ERROR:      File "/usr/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
ERROR:        return fut.result()
ERROR:    asyncio.exceptions.CancelledError
ERROR:
ERROR:    The above exception was the direct cause of the following exception:
ERROR:
ERROR:    Traceback (most recent call last):
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/uvicorn/protocols/http/h11_impl.py",
line 407, in run_asgi
ERROR:        result = await app(  # type: ignore[func-returns-value]
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/uvicorn/middleware/proxy_headers.py",
line 69, in __call__
ERROR:        return await self.app(scope, receive, send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/fastapi/applications.py", line 1054, in
__call__
ERROR:        await super().__call__(scope, receive, send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/applications.py", line 123,
in __call__
ERROR:        await self.middleware_stack(scope, receive, send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/middleware/errors.py", line
186, in __call__
ERROR:        raise exc
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/middleware/errors.py", line
164, in __call__
ERROR:        await self.app(scope, receive, _send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/middleware/cors.py", line 83,
in __call__
ERROR:        await self.app(scope, receive, send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/middleware/exceptions.py",
line 62, in __call__
ERROR:        await wrap_app_handling_exceptions(self.app, conn)(scope, receive,
send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line
64, in wrapped_app
ERROR:        raise exc
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line
53, in wrapped_app
ERROR:        await app(scope, receive, sender)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 758, in __call__
ERROR:        await self.middleware_stack(scope, receive, send)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 778, in app
ERROR:        await route.handle(scope, receive, send)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 299, in handle
ERROR:        await self.app(scope, receive, send)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 79, in app
ERROR:        await wrap_app_handling_exceptions(app, request)(scope, receive,
send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line
64, in wrapped_app
ERROR:        raise exc
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line
53, in wrapped_app
ERROR:        await app(scope, receive, sender)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 74, in app
ERROR:        response = await func(request)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/fastapi/routing.py",
line 278, in app
ERROR:        raw_response = await run_endpoint_function(
ERROR:      File "/usr/local/lib/python3.10/dist-packages/fastapi/routing.py",
line 191, in run_endpoint_function
ERROR:        return await dependant.call(**values)
ERROR:      File "/app/aphrodite-engine/aphrodite/endpoints/aws/api_server.py",
line 174, in create_completion
ERROR:        generator = await openai_serving_completion.create_completion(
ERROR:      File
"/app/aphrodite-engine/aphrodite/endpoints/openai/serving_completions.py", line
177, in create_completion
ERROR:        async for i, res in result_generator:
ERROR:      File
"/app/aphrodite-engine/aphrodite/endpoints/openai/serving_completions.py", line
81, in consumer
ERROR:        raise item
ERROR:      File
"/app/aphrodite-engine/aphrodite/endpoints/openai/serving_completions.py", line
66, in producer
ERROR:        async for item in iterator:
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 625, in generate
ERROR:        raise e
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 619, in generate
ERROR:        async for request_output in stream:
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 72, in __anext__
ERROR:        raise result
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 33, in _raise_exception_on_finish
ERROR:        task.result()
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 468, in run_engine_loop
ERROR:        has_requests_in_progress = await asyncio.wait_for(
ERROR:      File "/usr/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
ERROR:        raise exceptions.TimeoutError() from exc
ERROR:    asyncio.exceptions.TimeoutError
ERROR:
ERROR:    The above exception was the direct cause of the following exception:
ERROR:
ERROR:    Traceback (most recent call last):
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/uvicorn/protocols/http/h11_impl.py",
line 407, in run_asgi
ERROR:        result = await app(  # type: ignore[func-returns-value]
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/uvicorn/middleware/proxy_headers.py",
line 69, in __call__
ERROR:        return await self.app(scope, receive, send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/fastapi/applications.py", line 1054, in
__call__
ERROR:        await super().__call__(scope, receive, send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/applications.py", line 123,
in __call__
ERROR:        await self.middleware_stack(scope, receive, send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/middleware/errors.py", line
186, in __call__
ERROR:        raise exc
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/middleware/errors.py", line
164, in __call__
ERROR:        await self.app(scope, receive, _send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/middleware/cors.py", line 83,
in __call__
ERROR:        await self.app(scope, receive, send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/middleware/exceptions.py",
line 62, in __call__
ERROR:        await wrap_app_handling_exceptions(self.app, conn)(scope, receive,
send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line
64, in wrapped_app
ERROR:        raise exc
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line
53, in wrapped_app
ERROR:        await app(scope, receive, sender)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 758, in __call__
ERROR:        await self.middleware_stack(scope, receive, send)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 778, in app
ERROR:        await route.handle(scope, receive, send)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 299, in handle
ERROR:        await self.app(scope, receive, send)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 79, in app
ERROR:        await wrap_app_handling_exceptions(app, request)(scope, receive,
send)
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line
64, in wrapped_app
ERROR:        raise exc
ERROR:      File
"/usr/local/lib/python3.10/dist-packages/starlette/_exception_handler.py", line
53, in wrapped_app
ERROR:        await app(scope, receive, sender)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/starlette/routing.py",
line 74, in app
ERROR:        response = await func(request)
ERROR:      File "/usr/local/lib/python3.10/dist-packages/fastapi/routing.py",
line 278, in app
ERROR:        raw_response = await run_endpoint_function(
ERROR:      File "/usr/local/lib/python3.10/dist-packages/fastapi/routing.py",
line 191, in run_endpoint_function
ERROR:        return await dependant.call(**values)
ERROR:      File "/app/aphrodite-engine/aphrodite/endpoints/aws/api_server.py",
line 174, in create_completion
ERROR:        generator = await openai_serving_completion.create_completion(
ERROR:      File
"/app/aphrodite-engine/aphrodite/endpoints/openai/serving_completions.py", line
177, in create_completion
ERROR:        async for i, res in result_generator:
ERROR:      File
"/app/aphrodite-engine/aphrodite/endpoints/openai/serving_completions.py", line
81, in consumer
ERROR:        raise item
ERROR:      File
"/app/aphrodite-engine/aphrodite/endpoints/openai/serving_completions.py", line
66, in producer
ERROR:        async for item in iterator:
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 625, in generate
ERROR:        raise e
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 612, in generate
ERROR:        stream = await self.add_request(request_id,
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 503, in add_request
ERROR:        self.start_background_loop()
ERROR:      File "/app/aphrodite-engine/aphrodite/engine/async_aphrodite.py",
line 379, in start_background_loop
ERROR:        raise AsyncEngineDeadError(
ERROR:    aphrodite.engine.async_aphrodite.AsyncEngineDeadError: Background loop
has errored already.
INFO:     169.254.178.2:35944 - "GET /health HTTP/1.1" 200
```
The annoying part is that the server is not stopped, and the "health" still shows 200 (should not be the case, since the backend crashed).

**Comment by mrseeker** [comment]
If #265 fixes this issue, let me know, and then I will be happy to test it out.

**Comment by AlpinDale** [comment]
Seems like a timeout error. Did you have a sequence that took longer than 60 seconds to process? As a hotfix, you can increase the timeout threshold:

```sh
export APHRODITE_ENGINE_ITERATION_TIMEOUT_S=120
```

That would set the limit to 120 seconds. You can of course pass it as an env variable to the docker image.

**Comment by mrseeker** [comment]
Currently running on AWS. My setup:
- 8 GPU's
- 25 clients
- 1 server

And it crashes on a 60sec timeout, meaning it hogs a single GPU instead of distributing the load.

**Comment by mrseeker** [comment]
Normal response time for 1 GPU with 1 client is around 6 seconds.

**Comment by AlpinDale** [comment]
Most issues fixed with v0.6.0
